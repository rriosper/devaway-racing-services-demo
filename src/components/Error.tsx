import styled from '@xstyled/styled-components';
import { th } from '@xstyled/system';
import PropTypes, { InferPropTypes } from 'prop-types';
import React from 'react';

import { Card } from './Card';

const propTypes = {
  onClose: PropTypes.func,
  onRetry: PropTypes.func,
  message: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
};

const defaultProps = {
  onClose: null,
  onRetry: null,
};

const SButton = styled.button`
  outline: none;
  border: none;
  border-radius: onlyTop;
  cursor: pointer;
  transition: fast;
  border: thin;
  width: fit-content;
  padding-left: md;
  padding-top: sm;
  padding-bottom: sm;
  padding-right: md;
  border-color: foregroundBorder;
  color: foreground;

  :hover {
    background-color: foreground;
    border-color: foreground;
    color: white;
  }
`;

const SBody = styled.div`
  background-color: error;
  padding: md;
  color: white;
  border-radius: card;

  > p:first-child {
    font-weight: bold;
    font-size: 3;
  }
`;

const SActions = styled.div`
  display: flex;
  justify-content: end;
  padding: sm;
  gap: ${th('space.md')};
`;

const Error: React.FC<
InferPropTypes<typeof propTypes, typeof defaultProps>
> = ({
  onClose, onRetry, message, title,
}) => (
  <Card padding="none">
    <SBody>
      <p>{title}</p>
      <p>{message}</p>
    </SBody>
    <SActions>
      {onRetry && (
        <SButton
          onClick={() => {
            onRetry();
          }}
        >
          Retry
        </SButton>
      )}
      {onClose && <SButton onClick={onClose}>Close</SButton>}
    </SActions>
  </Card>
);

Error.propTypes = propTypes;
Error.defaultProps = defaultProps;

export default Error;
