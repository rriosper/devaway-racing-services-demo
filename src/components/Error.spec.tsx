import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from '@xstyled/styled-components';
import React from 'react';

import Error from './Error';

test('Error with retry and close', () => {
  const onRetry = jest.fn();
  const onClose = jest.fn();

  const { asFragment, getByText } = render(
    <ThemeProvider
      theme={{
        space: {
          md: '1px',
        },
      }}
    >
      <Error
        title="Error title"
        message="Error message"
        onRetry={onRetry}
        onClose={onClose}
      />
    </ThemeProvider>,
  );

  expect(getByText('Error title')).toBeInTheDocument();
  expect(getByText('Error message')).toBeInTheDocument();

  fireEvent.click(getByText('Close'));
  expect(onClose).toHaveBeenCalledTimes(1);

  fireEvent.click(getByText('Retry'));
  expect(onRetry).toHaveBeenCalledTimes(1);

  expect(asFragment).toMatchSnapshot();
});

test('Error only with close', () => {
  const onClose = jest.fn();

  const { getByText, queryByText } = render(
    <ThemeProvider
      theme={{
        space: {
          md: '1px',
        },
      }}
    >
      <Error title="Error title" message="Error message" onClose={onClose} />
    </ThemeProvider>,
  );

  expect(getByText('Error title')).toBeInTheDocument();
  expect(getByText('Error message')).toBeInTheDocument();

  fireEvent.click(getByText('Close'));
  expect(onClose).toHaveBeenCalledTimes(1);

  expect(queryByText('Retry')).not.toBeInTheDocument();
});

test('Error only with retry', () => {
  const onRetry = jest.fn();

  const { getByText, queryByText } = render(
    <ThemeProvider
      theme={{
        space: {
          md: '1px',
        },
      }}
    >
      <Error title="Error title" message="Error message" onRetry={onRetry} />
    </ThemeProvider>,
  );

  expect(getByText('Error title')).toBeInTheDocument();
  expect(getByText('Error message')).toBeInTheDocument();

  fireEvent.click(getByText('Retry'));
  expect(onRetry).toHaveBeenCalledTimes(1);

  expect(queryByText('Close')).not.toBeInTheDocument();
});
