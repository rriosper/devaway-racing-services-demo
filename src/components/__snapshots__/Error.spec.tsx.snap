// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error with retry and close 1`] = `[Function]`;
